packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -792,6 +804,7 @@ function dc(
     // to identify different constructors as vnodes with different keys to avoid reusing the
     // element used for previous constructors.
     data.key = `dc:${idx}:${data.key}`;
+    data.props = { ...data.props, ...props };
```

Comment:
I would recommend only doing this in the `isConfig` branch. This would avoid spreading cloning the `data.props` when not needed.